src/protocols/secure_channel/PASESession.cpp

```diff
@@ -523,52 +599,49 @@ CHIP_ERROR PASESession::HandleMsg1_and_SendMsg2(const System::PacketBufferHandle
     uint8_t verifier[kMAX_Hash_Length];
     size_t verifier_len = kMAX_Hash_Length;
 
-    uint16_t data_len; // To be initialized once we compute it.
-
-    const uint8_t * buf = msg->Start();
-    size_t buf_len      = msg->DataLength();
-
-    uint16_t encryptionKeyId = 0;
-
     ChipLogDetail(SecureChannel, "Received spake2p msg1");
 
-    VerifyOrExit(buf != nullptr, err = CHIP_ERROR_MESSAGE_INCOMPLETE);
-    VerifyOrExit(buf_len == sizeof(encryptionKeyId) + kMAX_Point_Length, err = CHIP_ERROR_INVALID_MESSAGE_LENGTH);
+    System::PacketBufferTLVReader tlvReader;
+    TLV::TLVType containerType = TLV::kTLVType_Structure;
 
-    err = mSpake2p.BeginVerifier(nullptr, 0, nullptr, 0, mPASEVerifier.mW0, kSpake2p_WS_Length, mPoint, sizeof(mPoint));
-    SuccessOrExit(err);
+    const uint8_t * X;
+    size_t X_len = 0;
+
+    tlvReader.Init(std::move(msg));
+    SuccessOrExit(err = tlvReader.Next(containerType, TLV::AnonymousTag));
+    SuccessOrExit(err = tlvReader.EnterContainer(containerType));
 
-    encryptionKeyId = chip::Encoding::LittleEndian::Read16(buf);
-    msg->ConsumeHead(sizeof(encryptionKeyId));
+    SuccessOrExit(err = tlvReader.Next());
+    VerifyOrExit(TLV::TagNumFromTag(tlvReader.GetTag()) == 1, err = CHIP_ERROR_INVALID_TLV_TAG);
+    X_len = tlvReader.GetLength();
+    SuccessOrExit(err = tlvReader.GetDataPtr(X));
+    SuccessOrExit(
+        err = mSpake2p.BeginVerifier(nullptr, 0, nullptr, 0, mPASEVerifier.mW0, kSpake2p_WS_Length, mPoint, sizeof(mPoint)));
 
     // Pass Pa to check abort condition.
-    err = mSpake2p.ComputeRoundOne(msg->Start(), msg->DataLength(), Y, &Y_len);
-    SuccessOrExit(err);
+    SuccessOrExit(err = mSpake2p.ComputeRoundOne(X, X_len, Y, &Y_len));
```

Comment:
BTW, what is called X here is called `pA` in spec since in spake2+ it says `then transmits pA=X to B`. The crypto output is called `X`, the message body is called `pA`.

Reply:
Updated based on this comment: https://github.com/project-chip/connectedhomeip/pull/9582#discussion_r706209284